power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -598,17 +598,88 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
 
+  public:
+    struct BinarySearch {
+        IntS lower_bound = -12;      // tap_position_left
+        IntS upper_bound = 12;       // tap_position_right
+        IntS current = 0;            // tap_position_middle
+        bool last_down = false;      // last direction
+        bool last_check = false;     // last check
+        bool tap_reverse = false;    // tap direction
+        bool inevitable_run = false; // inevitable run
+
+        BinarySearch() = default;
+        BinarySearch(IntS tap_pos, IntS tap_min, IntS tap_max) { reset_bs(tap_pos, tap_min, tap_max); }
+
+        void reset_bs(IntS tap_pos, IntS tap_min, IntS tap_max) {
+            last_down = false;
+            last_check = false;
+            current = tap_pos;
+            inevitable_run = false;
+            lower_bound = std::min(tap_min, tap_max);
+            upper_bound = std::max(tap_min, tap_max);
+            tap_reverse = tap_max < tap_min;
+        }
+
+        IntS get_bs_tap_left() const { return lower_bound; }
+        IntS get_bs_tap_right() const { return upper_bound; }
+        IntS get_bs_current_tap() const { return current; }
+        bool get_bs_last_down() const { return last_down; }
+        bool get_bs_last_check() const { return last_check; }
+        bool get_bs_tap_reverse() const { return tap_reverse; }
+        bool get_bs_inevitable_run() const { return inevitable_run; }
+        bool get_end_of_bs() const {
+            return get_bs_tap_left() >= get_bs_tap_right();
+            return (get_bs_tap_left() > get_bs_tap_right()) != get_bs_tap_reverse();
+        }
+
+        void set_bs_tap_left(IntS tap_left) { this->lower_bound = tap_left; }
+        void set_bs_tap_right(IntS tap_right) { this->upper_bound = tap_right; }
+        void set_bs_current_tap(IntS current_tap) { this->current = current_tap; }
+        void set_bs_last_down(bool last_down) { this->last_down = last_down; }
+        void set_bs_last_check(bool last_check) { this->last_check = last_check; }
+        void set_bs_tap_reverse(bool tap_reverse) { this->tap_reverse = tap_reverse; }
+        void set_bs_inevitable_run(bool inevitable_run) { this->inevitable_run = inevitable_run; }
+
+        bool adjust_bs(bool strategy_max = true) {
+            if (get_bs_last_down()) {
+                set_bs_tap_right(get_bs_current_tap());
+            } else {
+                set_bs_tap_left(get_bs_current_tap());
+            }
+            if (get_bs_tap_left() < get_bs_tap_right()) {
+                bool _max = strategy_max ? !get_bs_tap_reverse() : get_bs_tap_reverse();
+                IntS tap_pos = search_bs(_max);
+                if (get_bs_current_tap() == tap_pos) {
+                    return false;
+                }
+                set_bs_current_tap(tap_pos);
+                return true;
+            }
+            return false;
+        }
+        IntS search_bs(bool strategy_max = true) const {
```

Comment:
Since the interpretation of strategy_max here is different from the one in adjust_bs, maybe this can be renamed.

Reply:
In a way yes indeed.

Reply:
Comment addressed in the latest commit.